### PR TITLE
[3D] Geolocation

### PR DESCRIPTION
**Pull request by oterral**
[Test](https://mf-geoadmin3.dev.bgdi.ch/teo_geoloc/index.html?topic=ech&lang=it&bgLayer=ch.swisstopo.pixelkarte-farbe&dev3d=true&layers_visibility=false,false,false,false&geolocation=true&layers=ch.swisstopo.zeitreihen,ch.bfs.gebaeude_wohnungs_register,ch.bafu.wrz-wildruhezonen_portal,ch.swisstopo.swisstlm3d-wanderwege&layers_timestamp=18641231,,,&lon=7.43000&lat=46.24000&elevation=3000&heading=360.000&pitch=-90.000)

### CODE REVIEW

**Comment by cedricmoullet** [comment]
@gberaudo is it possible to update cesium in order to use the IOS9 fix they have in the last release ?

**Comment by oterral** [comment]
Altitude added

**Comment by oterral** [comment]
This PR needs review @gjn, @fredj, @gberaudo

File: src/components/geolocation/GeolocationDirective.js [hunk 1]
```diff
@@ -175,22 +126,31 @@ goog.require('ga_throttle_service');
...
               updateHeadingFeature(0);
-              mapHeadingRendering = false;
             }
           }
         };
 
+        var getCoords3d = function() {
+          var coords = geolocation.getPosition();
+          if (geolocation.getAltitude()) {
+            coords = coords.concat(1800);
```
Comment (gjn) [inline src/components/geolocation/GeolocationDirective.js]: What's this magic number?

File: src/components/geolocation/GeolocationDirective.js [hunk 2]
```diff
@@ -246,15 +206,17 @@ goog.require('ga_throttle_service');
...
-            updateUserTakesControl);
+        map.on('pointerdrag', updateUserTakesControl);
 
+        // Re-init userTakesControl when switching 2d/3d
+        scope.$watch(function() {
+            return scope.ol3d && scope.ol3d.getEnabled();
+        }, function(active) {
+          userTakesControl = false;
+          locate();
+        });
```
Comment (gjn) [inline src/components/geolocation/GeolocationDirective.js]: Can we unreg this when geolocation is not active?

**Comment by gjn** [comment]
@davidoesch @pfanguin Can you test this in the field? Both in 2D and in 3D.

File: src/components/geolocation/GeolocationDirective.js (hunk 1 above)
Comment (oterral) [inline src/components/geolocation/GeolocationDirective.js]: a relic of test

File: src/components/geolocation/GeolocationDirective.js (hunk 2 above)
Comment (oterral) [inline src/components/geolocation/GeolocationDirective.js]: yes maybe . I look.

**Comment by davidoesch** [comment]
if geolocation is on: pitch  -40

( a pity I cant test on ios9)

**Comment by gberaudo** [comment]
@cedricmoullet, I merged Cesium 1.14 in the c2c_patches branch; we can update dependency in the Makefile to latest commit.

**Comment by cedricmoullet** [comment]
thanks !

2015-10-09 13:39 GMT+02:00 gberaudo <notifications@github.com>:

> @cedricmoullet <https://github.com/cedricmoullet>, I merged Cesium 1.14
> in the c2c_patches branch; we can update dependency in the Makefile to
> latest commit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2706#issuecomment-146840283>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://twitter.com/cedricmoullet>

**Comment by oterral** [comment]
@gjn PR modified according to your comments

**Comment by gjn** [comment]
Could you alos update the cesium version (as mentioned my @gberaudo ?

**Comment by oterral** [comment]
It doesn't change anything for this PR. I'll made it in another PR.

**Comment by gjn** [comment]
ok.

Code is ok for me. I would like some outdoor tests with this, both in 2D and 3D.

**Comment by oterral** [comment]
Altitude seems to work but we have a problem with the heading. Dont' think is relevant in 3d to see an arrow which moves on 2 dimensions.

**Comment by gjn** [comment]
If 2D tests were successfull, please merge (maybe squash first).

**Pull request merged**